### ISSUE TITLE
Wrong column offset

### ISSUE THREAD

**Issue by albertosantini** [AtomLinter/linter-eslint#276]
In `5.1.0` there is something weird in column offset.

```
"use strict";

var line = 1

console.log(line);

(function (module) {

}());
```

```
λ eslint foo.js

C:\My\Dev\snippets\angular2-es5-seed\foo.js
  3:13  error  Missing semicolon                   semi
  7:12  error  "module" is defined but never used  no-unused-vars

✖ 2 problems (2 errors, 0 warnings)
```

In the panel `eslint-linter` reports
```
ESLint Error semi Missing semicolon.at line 3 col 12
ESLint Error no-unused-vars "module" is defined but never usedat line 7 col 12
```

Notice `line  3 col 12` vs. `3:13`; while the column of the other error is correct.

**Comment by steelbrain** [comment]
turns out, it's weirdly inaccurate. Works for some rules and not for others
![screenshot](https://cloud.githubusercontent.com/assets/4278113/11170391/fe5dcba8-8b8f-11e5-9bb7-a8ae76742550.png)

![screenshot](https://cloud.githubusercontent.com/assets/4278113/11170394/143de200-8b90-11e5-83e0-ffb4b4f0725c.png)

**Comment by albertosantini** [comment]
It seems the problem is only with the issues where the column error is the last char of the line.
I am preparing a PR. 

However it should be better handled in `rangeFromLineNumber`, because the logic is quite weak there, in my humble opionion.

**Comment by steelbrain** [comment]
I would love to accept your contributions to not just this project but to the other one as well. Nothing makes me happier than seeing people collaborate on projects, but it would really help me understand the matter if you would explain what you mean by weak logic :)

**Comment by albertosantini** [comment]
A few notes about `rangeFromLineNumber`. 

If column parameter is not passed, the end column is zero.
This value is not obvious in the code (and there is not documentation).

The column value is zero-based, so it is subtracted one to the value when the range is created,
but the linters use a logic one-based. Indeed they take the start column and add one, for instance, when they need to display info in the panel.

So for weak, I mean the render logic (zero-based, I suppose for underline tasks) is mixed with the editor one (one-based, column starts from one in the status bar). 

Of course, I may be very wrong and it is only an opinion. :)

**Comment by Arcanemagus** [comment]
@albertosantini the `colStart` parameter has nothing to do with the end column, that is derived from `textEditor.getBuffer().lineLengthForRow(lineNumber)`.

I think the step you are missing here is you are only thinking of `eslint`, the `linter` package consumes _many_ different linters, some of which are 0 based, some of which are 1 based. The `linter` package just has standardized their API on 0 based ranges, leaving it up to the service providers (like `linter-eslint`) to translate the ranges provided by the programs they wrap into ranges that `linter` understands.

**Comment by albertosantini** [comment]
@Arcanemagus thanks for the details.
I beg your pardon: no column based flame. :)

I would rephrase my opinion saying (maybe!) there is room for improvements there:
- using array of objects with explicit property, for instance `[{colStart: 11, colEnd: 17}, {colStart: 21, colEnd: 26}]` - but I am sure there are other APIs consuming an array of arrays.
- checking if `colStart > colEnd` to avoid `if (range[0][1] > range[1][1]) {` in the linter.
- passing `columnBased` argument, default to zero, to avoid column tweaks.

Just my two cents. Apologize me if I have been missing other details.

And maybe I should write these suggestions to `atom-linter` project. :)

For the sake of completeness you may give a look at this eslint issue:
https://github.com/eslint/eslint/issues/2284

**Comment by steelbrain** [comment]
@albertosantini The second point you made is very valid. However the I am not sure what we can do about the first and last ones. The ranges we're dealing with are not interpreted by us, but they come from the Atom's [`text-buffer` module](https://github.com/atom/text-buffer/blob/master/src/range.coffee). It's a feature of Atom API that every method that accepts a Range object, also accepts a range-like array.

  Therefore, the core does not allow us to use a range API like [{colStart: 11, colEnd: 17}, {colStart: 21, colEnd: 26}]`. About the third point you made, All of the lines and columns in Atom's `text-buffer` are `0-based`, so we had two options, either reduce the 1 index in the base linter or make all of the providers return 0-based indices. We went with the latter to be consistent with the core APIs because they don't allow 1-based indices. We have however allowed a `baseReduction` parameter in the `parse` method.

Anyways, I would to accept a PR or fix Point 2, It would be very nice of you if you would create a tracking issue for it in the [atom-linter](https://github.com/atomlinter/atom-linter) repo.

**Comment by albertosantini** [comment]
I studied again `rangeFromLineNumber` and it is ok.
I was wrong. Thanks again for the details.

The patch makes more sense if I don't use the `colStart` parameter.

```
            const range = Helpers.rangeFromLineNumber(textEditor, line - 1)
            if (column) {
              range[0][1] = column - 1
            }
            if (column > range[1][1]) {
                range[1][1] = column - 1
            }
```

Indeed the range, with zero-based column, is returned.

Then, if the error is in the column past the last char (equals to the length of the line + 1), we need to fix accordingly `colEnd` (`range[1][1]`).

About my point 2 in atom-linter. 
That control makes sense only when the programmer sets colStart to a wrong value (as I did - column), that's a value greater than the length of the line.

If you agree, I would prepare two PRs: one here and one in atom-linter.
Sorry for the noise.
